man/mosmlc.1
.TH MOSMLC 1 "18 September 2001" "Version 2.0"
.SH NAME
mosmlc \- Standard ML batch compiler
.SH SYNOPSIS
.B mosmlc [\fIOPTION\fR]... [\fIFILE\fR]...
.SH DESCRIPTION
.PP
Invokes the Moscow ML batch compiler and linker.  The compiler
compiles units and links programs, and can turn them into standalone
executable bytecode files.
.PP
The 
.B mosmlc
command has a command-line interface similar to that of most C
compilers.  It accepts several types of arguments: source files for
unit interfaces, source files for unit implementations, compiled unit
interfaces (which are added to the compilation context), and compiled
unit implementations (which are added to the linked executable).
.PP
A \fIFILE\fR argument ending in \fB.sig\fR is taken to be the name of
a source file containing a unit interface.  Given a file \fBU.sig\fR,
the compiler produces a compiled interface in the file \fBU.ui\fR.
.PP
A \fIFILE\fR argument ending in \fB.sml\fR is taken to be the name of
a source file containing a unit implementation.  Given a file \fBU.sml\fR,
the compiler produces compiled object code in the file \fBU.uo\fR. It
also produces an inferred interface file \fBU.ui\fR if there is no
explicit interface \fBU.sig\fR.
.PP
A \fIFILE\fR argument ending in \fB.ui\fR is taken to be the name of a
compilation unit interface.  The contents of that compilation unit
are added to the compilation context of the remaining source files.
.PP  
A \fIFILE\fR argument ending in \fB.uo\fR is taken to be the name of a
compiled unit implementation.  Such files are linked together, along
with the compiled unit implementations obtained by compiling \fB.sml\fR
arguments (if any), and the necessary Moscow ML library files, to produce
a standalone executable program.
.PP
The output of the linking phase is a file containing compiled code
that can be executed by the runtime system 
.B camlrunm.  
If \fBmosmlout\fR is the name of the file produced by the linking
phase (with option \fB\-o mosmlout\fR), the command
.B mosmlout arg1 ... argn
executes the compiled code contained in \fBmosmlout\fR.
.PP
Executing the code means executing the toplevel declarations of all
the bytecode files involved, in the order in which they were linked.
The list of command-line arguments \fBarg1\fR ... \fBargn\fR can
be obtained in a program by \fBCommandLine.arguments ()\fR.
.PP
There is no distinguished function which is automatically invoked when
the program is executed, but it is common to define a \fBmain\fR
function and invoke it using the toplevel declaration 
\fBval \_ = main()\fR.
.PP
This manual page documents the options accepted by 
.B mosmlc.
For more information consult the online manuals.
.SH OPTIONS
.TP
\fB\-c\fR
Compile only. Suppresses the linking phase of the compilation.
Source code files are turned into compiled files (\fB.ui\fR
and \fB.uo\fR), but no executable file is produced.  This option is
useful for compiling separate units.
.TP
\fB\-conservative\fR
Sets conservative mode for compilation of subsequent units: accept
all extensions to the SML Modules language, but issue a warning for
each use. This is the default.
.TP
\fB\-files\fR \fIresponse-file\fR
Pass the names of files listed in file \fIresponse-file\fR to the
linking phase just as if these names appeared on the command line.
File names in \fIresponse-file\fR are separated by blanks (spaces,
tabs, newlines) and must end either in \fB.sml\fR or \fB.uo\fR. A
name \fBU.sml\fR appearing in the response file is equivalent to
\fBU.uo\fR.
.TP
\fB\-g\fR
This option causes some information about exception names to be
written at the end of the executable bytecode file.
.TP
\fB\-g\fR
Causes the compiler to print the inferred interface(s) of the unit
implementation(s) being compiled.  Also causes the linker to list
all object files linked.  A \fBU.sig\fR file corresponding to a given
\fBU.sml\fR file can be produced semi-automatically by piping the
output of the compiler to a file \fBU.out\fR, and subsequently editing
this file to obtain a file \fBU.sig\fR.
.TP
\fB\-I\fR \fIdirectory\fR
Add the given directory to the list of directories searched for
compiled interface files (\fB.ui\fR) and compiled implementation
files (\fB.uo\fR).  By default, the current directory is searched
first, then the standard library directory.  Directories added with
\fB\-I\fR are searched after the current directory, but before the
standard library directory.  When several directories are added with
several \fB\-I\fR options on the command line, these directories are
searched from left to right.
.TP
\fB\-imptypes\fR
Specifies that the type checker should distinguish between
imperative and applicative type variables, generalize all
applicative type variables, and generalize imperative type variables
only in non-expansive expressions.
.TP
\fB\-liberal\fR
Sets liberal mode for compilation of subsequent units: accept
without warnings all extensions to the SML Modules language.
.TP
\fB\-msgstyle\fR \fIstyle\fR
By specifying \fB\-msgstyle msdev\fR, one can make the compiler
generate error messages understandable by Microsoft Developer
Studio.  The default behaviour is to generate error messages
understandable the Emacs editor in SML mode.
.TP
\fB\-noautolink\fR
The linker automatically links in any additional object files
required by the files explicitly specified on the command line.
With option \fB\-noautolink\fR all required object files must be
explicitly specified in the appropriate order.
.TP
\fB\-noheader\fR
Causes the output file produced by the linker to contain only the
bytecode, not preceded by any executable code.  A file 
\fBmosmlout\fR thus obtained can be executed only by explicitly
invoking the runtime system as follows: 
.B camlrunm mosmlout.
.TP
\fB\-o\fR \fIexec-file\fR
Specify the name of the output file produced by the linker.  In the
absence of this option, a default name, \fBa.out\fR is used.
.TP
\fB\-orthodox\fR
Sets orthodox mode for the compilation of subsequent units: reject
all uses of the extensions to the SML Modules language.  That is,
accept only SML Modules syntax.
.TP
\fB\-P\fR \fIunit-set\fR
Determines which library units will be open at compile-time.
Any library unit in the load path can be used by the compiler for
type checking purposes.  Thus regardless of the \fB\-P\fR option,
the compiler knows the type of library functions such as
\fBArray.foldl\fR.
.RS
.TP
\fB\-P\fR \fIdefault\fR
The units \fBChar\fR, \fBList\fR, and \fBString\fR will be partially
openend. This is the default, permitting e.g. \fBString.concat\fR
to be referred to just as \fBconcat\fR.
.TP
\fB\-P sml90\fR 
Provides an initial environment which is upwards compatible with
that of the 1990 `Definition of Standard ML' and with pre-1.30 releases
of Moscow ML. In particular, the functions \fBchr\fR, \fBexplode\fR,
\fBimplode\fR, and \fBord\fR work on strings, not characters.   The
math functions and input-output facilities required by the 1990
Definition are available at top-level.  In addition the same libraries
are opened as with \fB-P default\fR.
.TP
\fB\-P nj93\fR
Provides a top-level environment which is mostly compatible with
that of SML/NJ 0.93.  The functions \fBapp\fR, \fBceiling\fR,
\fBchr\fR, \fBdec\fR, \fBexplode\fR, \fBfold\fR, \fBhd\fR, \fBimplode\fR,
\fBinc\fR, \fBmax\fR, \fBmin\fR, \fBnth\fR, \fBnthtail\fR, \fBord\fR,
\fBordof\fR, \fBrevapp\fR, \fBrevfold\fR, \fBsubstring\fR, \fBtl\fR,
and \fBtruncate\fR have the same type and meaning as in SML/NJ 0.93.
Note that this is incompatible with SML/NJ version 110. The math
functions and input-output facilities required by the 1990 `Definition
of Standard ML' are available at top-level.  In addition the same (new)
libraries are opened as with \fB-P default\fR.  This option does not
imply \fB-imptypes\fR.
.TP
\fB\-P full\fR
Same as \fB\-P default\fR.
.TP
\fB\-P none\fR
No library units are initially opened.
.PP
Additional directories to be searched for library units can be
specified with the \fB-I\fR \fIdirectory\fR option.
.RE
.TP
\fB\-q\fR
Enables the quotation/antiquotation mechanism.
.TP
\fB\-standalone\fR
Specifies that the runtime system should be prepended to the
linked bytecode, thus creating a stand-alone executable.  This adds
75--100 KB to the size of the linked file. 
.TP
\fB\-stdlib\fR \fIstdlib-directory\fR
Specifies the standard library directory, which will be searched by
the compiler and linker for the \fB.ui\fR and \fB.uo\fR files
corresponding to units mentioned in the files being linked.  The
default standard library is set when the system is created; under
Debian it is \fB/usr/lib/mosml\fR.
.TP
\fB\-structure\fR
Specifies that subsequent \fB.sml\fR and \fB.sig\fR source files
must be compiled in `structure' mode.
.TP
\fB\-toplevel\fR
Specifies that subsequent \fB.sml\fR and \fB.sig\fR source files
must be compiled in `toplevel' mode.
.TP
\fB\-v\fR
Prints the version number of the various passes of the compiler.
.TP
\fB\-valuepoly\fR
Specifies that the type checker should use `value polymorphism'.
Default.
.SH AUTHOR
Compiled by Henning Niss from the various Moscow ML documents 
written by Sergei Romanenko, Claudio Russo, and Peter Sestoft.
.SH "SEE ALSO"
.BR mosml (1),
.BR mosmlyac (1),
.BR mosmllex (1)
.PP
The full documentation for
.B mosmlc
and the other Moscow ML tools can be found in the online manuals in 
\fB/usr/share/doc/mosml/manual.pdf.gz\fR
(`Moscow ML Owner's Manual') and
\fB/usr/share/doc/mosml/mosmllib.pdf.gz\fR
(`Moscow ML Library Documentation').


CHANGELOG.md
# Unreleased

* SML Basis Library changes:
  * Added the structures `PackRealBig` and `PackRealLittle`
  * Added stop-gap implementation of the structure `Int64` and
    `Word64`. Stop-gap meaning that they don't fully follow the Basis
    Library specification, in particular there is no support for
    overloaded constant literals nor overloaded operations.

* Fixed some bugs in the configure part of the build systems, which
  meant that some features was sometimes disabled

* Updated the various test-suites so that they can be used to easily
  check if the compiler and mosmllib works.


# Version 2.10.1 (August 2014)

* ML Server Pages added to examples

* Build system cleaned up a bit


# Version 2.10 (August 2013)

* Support for dynamic linking and callback from C also under MacOS X

* There may now be 2 G globals (string literals etc) instead of 64 K

* Bugs fixed:
  * The HTML files generated for mosmllib now have valid URLs
  * `Substring.app` more efficient
  * Recursive structure compilation (elabRecSigExp) fixed
  * Several errors in `Mosmlcookie`
  * Misplaced double quote in `Msp.ahrefa`
  * Exception `Option` wasn't available at top-level
  * Linker did not check stamps of linked-in units
  * Hash function now much faster on very long strings
  * Double alignment constraints now correct with gcc-3.2 and Solaris
  * Bug in `Polyhash.filter`
  * `Socket` library returned only first 16 bytes of UDP datagram

* SML Basis Library changes:
  * `String` and `Substring`: added `concatWith`, `isSuffix`,
    `isSubstring`, `full`
  * `ListPair`: added `zipEq`, `appEq`, `mapEq`, `foldlEq`, `foldrEq`,
    `allEq`
  * Added `ArraySlice` and `VectorSlice` structures, also for `Char` and `Word8`
  * Added `find`, `findi`, `all`, `exists`, `collate` to `vector`, `array` and
    `slice` structures
  * Added `collate` to `List`
  * Added `update` to vector structures
  * `Path.{mkRelative,mkAbsolute}` now take record arguments
  * Negative `Time.time` values allowed
  * `OS.Process`: new functions `sleep` and `isSuccess`; `status` not `eqtype`
  * `Listsort`: added `eqclasses`, `merge`, `mergeUniq`
  * `Path`: added functions `isRoot`, `fromUnixPath`, `toUnixPath`, and
    exception `InvalidArc` (bug report by Henning Niss)
  * `TextIO`: `inputLine` now have type `instream -> string option`
  * `Array`: added type abbreviation `vector` (bug report by Andrzej
    Wasowski)
  * `FileSys`: type `access` renamed to `access_mode` (bug report by
    Henning Niss)
  * `Byte`: `unpackString{,Vec}` now uses vector slices (bug report by
    Henning Niss)
  * `Unix`: added functions `fromStatus`, `textInstreamOf`, `binInstreamOf`,
    `textOutstreamOf`, `binOutstreamOf`, `exit`, and added phantom types
    to proc (bug report by Henning Niss)
  * `Timer`: added function `checkCPUTimes`
  * `Word` and `Word8`: added functions `toLarge`, `toLargeX`, and `fromLarge`
    (bug report by Martin Elsman)
  * `General`: added exception Span and made the type of the function
    `before` less general (bug report by Henning Niss)

* Added modules to mosmllib:
  * `Buffer`: mutable string buffers for fast and efficient
    concatenation of strings
  * `Hashset`: sets implemented by hash-tables
  * `Rbset`: ordered sets implemented by red-black trees
  * `Redblackmap`: maps implemented by red-black trees

* Moscow ML is now developed in the open at github.


# Version 2.00 (June 2000)

* The full SML Modules language (structures, signatures, and functors)
  is now supported, thanks to Claudio Russo.  Also, several extensions
  to the SML Modules language are provided:
      - higher-order functors: functors may be defined within structures
        and functors
      - first-class modules: structures and functors may be packed and
        then handled as Core language values, which may then be unpacked
        as structures or functors again
      - recursive modules: signatures and structures may be recursively
        defined

* Value polymorphism has become friendlier: non-generalizable free
  type variables are left free, and become instantiated (once only)
  when the bound variable is used

* Added facilities for creating and communicating with subprocesses
  (structure `Unix` and `Signal` from SML Basis Library).

* Added facilities for efficient functional generation of HTML code
  (structure `Msp`); also supports the writing of ML Server Page scripts.

* Added facilities setting and accessing 'cookies' in CGI scripts
  (structure `Mosmlcookie`), thanks to Hans Molin, Uppsala, Sweden.

* The `Gdimage` structure now produces PNG images (using Thomas
  Boutell's gd library).


# Version 1.44 (August 1999)

## Added

* Access to GNU gdbm persistent hashtables (structures `Gdbm`, `Polygdbm`)

* Interface to the PostgreSQL database server (structure `Postgres`)

* Interface to the MySQL database server (structure `Mysql`)

* Interface to POSIX 1003.2 regular expressions (structure `Regex`)

* Interface to sockets (structure `Socket`)

* Faster bytecode execution (when compiled with GCC or egcs)

* Registration of ML and C functions simplify callbacks


src/launch.w32/mosmlc.c
#include <windows.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <sys\stat.h>
#include "driver.h"

// -----

BOOL WINAPI MyHandlerRoutine( DWORD dwCtrlType )
{
  if( dwCtrlType == CTRL_C_EVENT )
    return TRUE;
  else
    return FALSE;
}

int ExecCmd(char *cmd_args)
{
  BOOL fOk;
  PROCESS_INFORMATION pi;
  STARTUPINFO si;
  DWORD dwExitCode;

  //fprintf(stderr, "\t%s\n", cmd_args);

  /* Initialize the STARTUPINFO structure: */

  ZeroMemory(&si, sizeof(si));
  si.cb = sizeof(si);

  /* Ignore CTRL-C. */
  fOk = SetConsoleCtrlHandler(MyHandlerRoutine, TRUE);
  Sleep(1);

  /* Start the shelled application: */
  fOk = CreateProcess(NULL, cmd_args, NULL, NULL, TRUE,
              NORMAL_PRIORITY_CLASS, NULL, NULL, &si, &pi);
  if( !fOk )
  {
    DWORD dwLastError;
    fprintf(stderr, "Unable to execute %s\n", cmd_args);
    dwLastError = GetLastError();
    switch( dwLastError )
    {
    default:
      fprintf(stderr, "Last Error = %x\n", dwLastError);
    }
    exit(2);
  }

  /* Wait for the shelled application to finish: */
  if( WaitForSingleObject(pi.hProcess, INFINITE) == WAIT_FAILED )
  {
    fprintf(stderr, "WaitForSingleObject failed\n");
    exit(2);
  }
  if( !GetExitCodeProcess(pi.hProcess, &dwExitCode) )
  {
    fprintf(stderr, "GetExitCodeProcess failed\n");
    exit(2);
  }
  CloseHandle(pi.hProcess);

  /* Restore the default CTRL-C handler. */
  fOk = SetConsoleCtrlHandler(NULL, FALSE);

  return dwExitCode;
}

// -----

char* camlrunm = "camlrunm.exe";
char* linkout = "mosmlout.exe";

main(int argc, char** argv)
{
  int i;
  char* a;
  extern char* getenv();
  char* cmd;
  int status;
  char* stdlib;

  int linkalso = 1;
  char* includes = NULL;
  char* compopt = NULL;
  char* linkopt = NULL;
  char* linkfiles = NULL;
  char* context = NULL;

  format(&includes, "");
  format(&compopt, "-conservative");
  format(&linkopt, "");
  format(&linkfiles, "");
  format(&context, "-structure");

  stdlib = getenv("MOSMLLIB");
  if( stdlib == NULL )
  {
    fprintf(stderr, "Variable MOSMLLIB is undefined.\n");
    exit(2);
  }

  for( i = 1; i < argc; i++ )
  {
    a = argv[i];
    if( suffix(a, ".sml") )
    {
      format(&cmd, "%s \"%s\\mosmlcmp\" -stdlib \"%s\" %s %s %s \"%s\"",
             camlrunm, stdlib, stdlib, includes, compopt, context, a);
      //fprintf(stderr, "%s\n", cmd);
      status = ExecCmd(cmd);
      if( status ) exit(status);
      free(cmd);
      reformat(&context, "%s \"%s.ui\"", context, woSuffix(a));
      reformat(&linkfiles, "%s \"%s\"", linkfiles, a);
    }
    else if( suffix(a, ".sig") )
    {
      format(&cmd, "%s \"%s\\mosmlcmp\" -stdlib \"%s\" %s %s %s \"%s\"",
             camlrunm, stdlib, stdlib, includes, compopt, context, a);
      //fprintf(stderr, "%s\n", cmd);
      status = ExecCmd(cmd);
      if( status ) exit(status);
      free(cmd);
      reformat(&context, "%s \"%s.ui\"", context, woSuffix(a));
    }
    else if( suffix(a, ".ui") )
    {
      reformat(&context, "%s \"%s\"", context, a);
    }
    else if( suffix(a, ".uo") )
    {
      reformat(&linkfiles, "%s \"%s\"", linkfiles, a);
    }
    else if( eq(a, "-structure") || eq(a, "-toplevel") )
    {
      reformat(&context, "%s %s", context, a);
    }
    else if( eq(a, "-c") )
    {
      linkalso = 0;
    }
    else if( eq(a, "-I") || eq(a, "-include") )
    {
      reformat(&includes, "%s -I \"%s\"", includes, argv[++i]);
    }
    else if( eq(a, "-P") || eq(a, "-perv") )
    {
      reformat(&compopt, "%s -P %s", compopt, argv[++i]);
      reformat(&linkopt, "%s -P %s", linkopt, argv[i]);
    }
    else if( eq(a, "-q") || eq(a, "-quotation") )
    {
      reformat(&compopt, "%s -q", compopt);
    }
    else if( eq(a, "-i") )
    {
      reformat(&compopt, "%s -i", compopt);
      reformat(&linkopt, "%s -i", linkopt);
    }
    else if( eq(a, "-g") || eq(a, "-debug") )
    {
      reformat(&compopt, "%s -g", compopt);
      reformat(&linkopt, "%s -g", linkopt);
    }
    else if( eq(a, "-m") || eq(a, "-msgstyle") )
    {
      reformat(&compopt, "%s -msgstyle %s", compopt, argv[++i]);
    }
    else if( eq(a, "-noheader") )
    {
      reformat(&linkopt, "%s -noheader", linkopt);
    }
    else if( eq(a, "-noautolink") )
    {
      reformat(&linkopt, "%s -noautolink", linkopt);
    }
    else if( eq(a, "-o") || eq(a, "-exec") )
    {
      linkout = argv[++i];
    }
    else if( eq(a, "-standalone") )
    {
      reformat(&linkopt, "%s -standalone", linkopt);
    }
    else if( eq(a, "-stdlib") )
    {
      stdlib = argv[++i];
    }
    else if( eq(a, "-v") || eq(a, "-version") )
    {
      printf("The Moscow ML system for the 80386 PC, version 2.00 (June 2000)\n");
      printf("  (standard library from %s)\n", stdlib);
      format(&cmd, "%s -V", camlrunm);
      ExecCmd(cmd);
      free(cmd);
      format(&cmd, "%s \"%s\\mosmlcmp\" -version", camlrunm, stdlib);
      ExecCmd(cmd);
      free(cmd);
      format(&cmd, "%s \"%s\\mosmllnk\" -version", camlrunm, stdlib);
      ExecCmd(cmd);
      free(cmd);
    }
    else if( eq(a, "-imptypes") )
    {
      reformat(&compopt, "%s -imptypes", compopt);
    }
    else if( eq(a, "-valuepoly") )
    {
      reformat(&compopt, "%s -valuepoly", compopt);
    }
    else if( eq(a, "-orthodox") || eq(a, "-conservative") || eq(a, "-liberal") )
    {
      reformat(&compopt, "%s %s", compopt, a);
    }
    else if( eq(a, "-files") )
    {
      reformat(&linkfiles, "%s -files \"%s\"", linkfiles, argv[++i]);
    }
    else if( prefix(a, "-") )
    {
      fprintf(stderr, "Unknown option \"%s\", ignored\n", a);
    }
    else
    {
      fprintf(stderr, "I don't know what to do with file \"%s\", ignored\n",
              a);
    }
  }

  if( linkalso && linkfiles[0] != 0 )
  {
    format(&cmd,
           "%s \"%s\\mosmllnk\" -stdlib \"%s\" %s %s -exec \"%s\" %s",
           camlrunm, stdlib, stdlib, includes,
           linkopt, linkout, linkfiles);
    // fprintf(stderr, "%s\n", cmd);
    status = ExecCmd(cmd);
    return status;
  }
  return 0;
}
